Improve code of class BePlusTree

hvlad opened the pull request:
Maked it usabe with debugger, removed a lot of casts and a few other misc improvemens.

hvlad committed:
Remove a lot of casts of void* to the ItemList* and NodeList*.
Make code friendly for debugging.

hvlad committed:
Replace NULL by nullptr, use uppercase for constants (LeafCount, NodeCount -> LEAF_COUNT, NODE_COUNT)

aafemt commented:
Access to a member of unions that wasn't recently assigned is UB in C++. For reliability it would be better to remove `void* pointer` from `NodePtr` and make explicit separate constructors for node pointer and item pointer.

aafemt commented on src/common/classes/tree.h:
```diff
@@ -1007,10 +1020,10 @@ void BePlusTree<Value, Key, Allocator, KeyOfValue, Cmp>::_removePage(const int n
 
 	if (nodeLevel)
-		((NodeList *)node)->~NodeList();
+		node.nodes->~NodeList();
```
`delete(pool) node.nodes` and `delete(pool) node.items` here and above, please.

hvlad committed:
Removed never used MallocAllocator and use common FB_NEW_POOL and delete instead of allocate/deallocate.
Removed excess "friend class" declarations.
NodePtr changed to avoid potential undefined behavior.

AlexPeshkoff commented on src/common/classes/tree.h:
```diff
@@ -106,24 +93,24 @@ enum LocType { locEqual, locLess, locGreat, locGreatEqual, locLessEqual };
 // more classical B+ tree feature). This is also not done to improve tree performance a little
 //
-template <typename Value, typename Key = Value, typename Allocator = MallocAllocator,
+template <typename Value, typename Key = Value, typename Allocator = Firebird::MemoryPool,
```
Why not remove "Allocator" parameter at all?

hvlad replied:
Good idea

hvlad replied:
Done

AlexPeshkoff commented on src/common/classes/tree.h:
```diff
@@ -106,24 +93,24 @@ enum LocType { locEqual, locLess, locGreat, locGreatEqual, locLessEqual };
 public:
 	explicit BePlusTree(Allocator *_pool)
-		: pool(_pool), level(0), root(NULL), defaultAccessor(this)
+		: pool(_pool), level(0), defaultAccessor(this)
```
Also suggest to use reference (not pointer) to MemoryPool in all c-tors - like in the rest of our classes.

hvlad replied:
Also OK

hvlad replied:
Done

hvlad committed:
Follow @AlexPeshkoff suggestions:
  removed Allocator from template arguments,
  removed pointer based ctor in favor of reference based one.

hvlad commented:
Should it be backported and, if yes, into which branches ?

dyemanov commented:
Usually, we don't backport refactoring commits. If you think it's really needed, I may agree with backporting into v5 only.

hvlad commented:
The main goal of this PR is not the refactoring. It is about debugging experience, which was close to impossible.

dyemanov commented:
OK, I don't mind (except maybe v3 which is nearly end-of-life). I just suppose that the need of debugging there may be over-estimated, given the last bugfix in this class was about 20 years ago. And even if we have something fixed in master/v5, it can be easily backported "as is".

hvlad commented:
Let me clarify - it is not about debugging `BePlusTree` only, it is mostly about debugging everything else where it is used.
I.e. before this patch it was hard to impossible to inspect tree or map contents.

AlexPeshkoff commented:
@hvlad +1
